### PR TITLE
implementation of the exponentiation operator

### PR DESCRIPTION
**Pull request by mekolat**
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

[//]: # (Note: Lines beginning with syntax such as this one, are comments and will not be visible in your report!)

### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [X] I have followed [proper Hercules code styling][code].
- [X] I have read and understood the [contribution guidelines][cont] before making this PR.
- [X] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
This is an implementation of the exponentiation operator (from #1667) with left-assignment.

* `pow(5, 3)` => `5 ** 3`
* `pow(pow(5, 3), 2)` => `5 ** 3 ** 2`
* `pow(5, pow(3, 2))` => `5 ** (3 ** 2)`
* `.@var = pow(.@var, 3)` =>`.@var **= 3`

The exponentiation operator has higher precedence than the multiplication operator and lower precedence than the unary operators. This means:

* `5 * 2 ** 3` is equal to `5 * (2 ** 3)`
* `5 ** 2 * 3` is equal to `(5 ** 2) * 3`

<br>

[//]: # (Describe at length, the changes that this pull request makes.)

**Affected Branches:** `master`
**Related issues:** #1667
**Type:** [enhancement](https://github.com/HerculesWS/Hercules/labels/type%3Aenhancement)
**Compopents:** [script engine](https://github.com/HerculesWS/Hercules/labels/component%3Acore%3Ascriptengine), [documentation](https://github.com/HerculesWS/Hercules/labels/component%3Adocumentation)

[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request if you have not already enabled it.)

---

### Questions
- Does the operator use the right level of precedence?
- Should the `pow()` buildin be marked as deprecated?
- Should existing scripts be converted so they use the exponentiation operator?

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/1739)

**Comment by Emistry** [comment]
`Should the pow() buildin be marked as deprecated?`
I believe not to mark it as deprecated.

Just like `Set( )` ... they still have their own use-case where the new/current operator cant work in these theircases...

**Comment by MishimaHaruna** [comment]
> Does the operator use the right level of precedence?

Yes, in mathematics, exponentiation takes precedence over multiplication, so I agree with the precedence you assigned.

The problem is that it seems to be generally recognized as being right-associative (see [wikipedia](https://en.wikipedia.org/wiki/Order_of_operations)), so:

```
a ** b ** c
```

should equal to

```
a ** (b ** c)
```

... and I don't remember having any support for this in the parser right now. (I wonder how troublesome it would be to add it, and whether it's worth the effort, or it's better to just keep it left-associative)

> Should the pow() buildin be marked as deprecated?

I think it should. I don't think we'd have cases where it's still necessary to use the pow() command once this is implemented, and I don't like having duplicate facilities.

> Should existing scripts be converted so they use the exponentiation operator?

Definitely. There shouldn't be many I believe.

**Comment by mekolat** [comment]
>The problem is that it seems to be generally recognized as being right-associative

I have no clue how I would make it right-associative without adding extraneous code... Is this a deal-breaker?

---

> I don't think we'd have cases where it's still necessary to use the pow() command once this is implemented, and I don't like having duplicate facilities.

Ok, I'll mark `pow` as deprecated; Is using `BUILDIN_DEF_DEPRECATED` enough or should I explicitly add a `DeprecationWarning()` to `pow` to tell the user to use the exponentiation operator?

---

> There shouldn't be many I believe.

Last time I checked there was only two scripts using it.

**Comment by MishimaHaruna** [comment]
Yes, let's keep it right associative, I don't have any ideas that wouldn't make parsing even uglier than it is. It's not deal-breaker, we can perhaps add a warning in the documentation, to make sure people will notice the left-associativity mention in the precedence table, even if they're used to other languages where this is right-associative.

File: db/re/item_db.conf [hunk 1]
```diff
@@ -6419,7 +6419,7 @@ item_db: (
 		bonus2 bIgnoreDefRate,RC_DemiPlayer,25;
 		bonus bUnbreakableWeapon,0;
 		if(getrefine()>5) {
-			bonus2 bAddRace,RC_DemiPlayer,pow(min(getrefine(), 14)-3,2);
+			bonus2 bAddRace,RC_DemiPlayer,((min(getrefine(), 14)-3) ** (2));
```
Comment (mekolat) [inline db/re/item_db.conf]: I was too lazy to go through the whole database so I used `sed`, hence the extraneous parentheses

**Comment by mekolat** [comment]
@MishimaHaruna if you don't need any other changes, I think this PR is ready to merge

**Pull request merged**